src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/unuseddeductions/service/UnusedDeductionsCalculationService.kt

```diff
@@ -0,0 +1,74 @@
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.AdjustmentServiceAdjustment
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.AdjustmentServiceAdjustmentType
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.BookingAdjustment
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.BookingAdjustmentType
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.BookingAndSentenceAdjustments
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.service.BookingService
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.service.CalculationService
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.service.PrisonService
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.validation.ValidationService
+import kotlin.math.max
+
+@Service
+class UnusedDeductionsCalculationService(
+  private val prisonService: PrisonService,
+  private val calculationService: CalculationService,
+  private val validationService: ValidationService,
+  private val bookingService: BookingService,
+) {
+
+  fun calculate(adjustments: List<AdjustmentServiceAdjustment>, offenderNo: String): Int? {
+    val prisoner = prisonService.getOffenderDetail(offenderNo)
+    val sourceData = prisonService.getPrisonApiSourceData(prisoner).copy(bookingAndSentenceAdjustments = useAdjustmentsFromAdjustmentsApi(adjustments))
+
+    val calculationUserInputs = CalculationUserInputs(useOffenceIndicators = true)
+
+    var validationMessages = validationService.validateBeforeCalculation(sourceData, calculationUserInputs)
+    if (validationMessages.isNotEmpty()) {
+      return null
+    }
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    validationMessages = validationService.validateBeforeCalculation(booking)
+    if (validationMessages.isNotEmpty()) {
+      return null
+    }
+
+    val result = calculationService.calculateReleaseDates(booking)
+    val releaseDateTypes = listOf(ReleaseDateType.CRD, ReleaseDateType.ARD, ReleaseDateType.MTD)
+    val calculationResult = result.second
+    var releaseDate = calculationResult.dates.filter { releaseDateTypes.contains(it.key) }.minOfOrNull { it.value }
+    val releaseDateSentence = result.first.getAllExtractableSentences().find { it.sentenceCalculation.releaseDate == releaseDate }
+    val unusedDeductions = if (releaseDateSentence != null) { // The release date will be a PRRD, no unused deductions.
+      val maxDeductions =
+        releaseDateSentence.sentenceCalculation.numberOfDaysToDeterminateReleaseDate
+
+      val remand = adjustments.filter { it.adjustmentType == AdjustmentServiceAdjustmentType.REMAND }.map { it.daysBetween!! }.reduceOrNull { acc, it -> acc + it } ?: 0
+      val taggedBail = adjustments.filter { it.adjustmentType == AdjustmentServiceAdjustmentType.TAGGED_BAIL }.map { it.days!! }.reduceOrNull { acc, it -> acc + it } ?: 0
```

Comment:
Are daysBetween and days guaranteed to not be null here? Would it make sense to filter out any null values first?

Reply:
Yes they've been filtered out in the previous class. They're only null if the record was created and synchronised from NOMIS.